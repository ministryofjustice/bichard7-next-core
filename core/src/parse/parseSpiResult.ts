import { XMLParser } from "fast-xml-parser"
import { decodeAttributeEntitiesProcessor, decodeTagEntitiesProcessor } from "src/lib/encoding"
import type { IncomingMessageParsedXml } from "../types/SpiResult"
import { incomingMessageParsedXmlSchema } from "../types/SpiResult"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeAttributeEntitiesProcessor,
    tagValueProcessor: decodeTagEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
