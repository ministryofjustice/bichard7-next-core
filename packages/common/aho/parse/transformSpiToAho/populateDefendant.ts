import type { Address, DefendantDetail, HearingDefendant } from "../../../types/AnnotatedHearingOutcome"
import type {
  ResultedCaseMessageParsedXml,
  SpiAddress,
  SpiCourtIndividualDefendant,
  SpiDefendant
} from "../../../types/SpiResult"

import { lookupRemandStatusBySpiCode } from "../../dataLookup/dataLookup"
import populateOffences from "./populateOffences"

const formatPncIdentifier = (spiPNCIdentifier?: string): string | undefined =>
  spiPNCIdentifier ? spiPNCIdentifier.substring(0, 4) + "/" + spiPNCIdentifier.substring(4) : undefined

const populatePersonDefendantDetail = (spiCourtIndividualDefendant: SpiCourtIndividualDefendant): DefendantDetail => {
  const {
    PersonDefendant: {
      BasePersonDetails: {
        Birthdate: spiBirthDate,
        Gender: spiGender,
        PersonName: {
          PersonFamilyName: spiPersonFamilyName,
          PersonGivenName1: spiPersonGivenName1,
          PersonGivenName2: spiPersonGivenName2,
          PersonGivenName3: spiPersonGivenName3,
          PersonTitle: spiPersonTitle
        }
      }
    }
  } = spiCourtIndividualDefendant

  const spiGivenNames = [spiPersonGivenName1, spiPersonGivenName2, spiPersonGivenName3]
    .filter((name) => name !== undefined && name.trim() !== "")
    .map((name) => name!.trim())
    .map((name) => name.replace(/\s+/g, " "))

  return {
    BirthDate: spiBirthDate ? new Date(spiBirthDate) : undefined,
    Gender: Number(spiGender),
    PersonName: {
      FamilyName: spiPersonFamilyName.replace(/\s+/g, " ").trim(),
      GivenName: spiGivenNames,
      Title: spiPersonTitle?.trim() || undefined
    }
  }
}

const populateAddress = (spiAddress: SpiAddress): Address => {
  if ("SimpleAddress" in spiAddress) {
    const {
      SimpleAddress: { AddressLine1, AddressLine2, AddressLine3, AddressLine4, AddressLine5 }
    } = spiAddress

    return {
      AddressLine1: AddressLine1,
      AddressLine2: AddressLine2,
      AddressLine3: AddressLine3,
      AddressLine4: AddressLine4,
      AddressLine5: AddressLine5
    }
  } else {
    const {
      ComplexAddress: { AdministrativeArea, Locality, StreetDescription, Town, UniqueStreetReferenceNumber }
    } = spiAddress

    const hearingOutcomeAddress = [
      UniqueStreetReferenceNumber?.toString(),
      StreetDescription,
      Locality,
      Town,
      AdministrativeArea
    ]
      .filter((x) => !!x)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .reduce((acc: any, addressPart, index) => {
        acc[`AddressLine${index + 1}`] = addressPart!
        return acc
      }, {})

    return hearingOutcomeAddress
  }
}

const parseBailConditions = (spiBailConditions?: string): string[] => {
  if (spiBailConditions && spiBailConditions !== "") {
    return spiBailConditions
      .split(";")
      .map((cond) => cond.trim())
      .filter((cond) => !!cond)
  }

  return []
}

type PartialDefendantDetails =
  | Pick<HearingDefendant, "Address" | "BailConditions" | "OrganisationName" | "PNCIdentifier" | "RemandStatus">
  | Pick<
      HearingDefendant,
      | "Address"
      | "BailConditions"
      | "CourtPNCIdentifier"
      | "DefendantDetail"
      | "ReasonForBailConditions"
      | "RemandStatus"
    >

const processDefendant = (spiDefendant: SpiDefendant): PartialDefendantDetails => {
  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        Address: spiAddress,
        BailStatus: spiBailStatus,
        PersonDefendant: { BailConditions: spiBailConditions, PNCidentifier: spiPNCidentifier },
        ReasonForBailConditionsOrCustody: spiReasonForBailConditionsOrCustody
      }
    } = spiDefendant

    return {
      Address: populateAddress(spiAddress),
      BailConditions: parseBailConditions(spiBailConditions),
      CourtPNCIdentifier: formatPncIdentifier(spiPNCidentifier),
      DefendantDetail: populatePersonDefendantDetail(spiDefendant.CourtIndividualDefendant),
      ReasonForBailConditions: spiReasonForBailConditionsOrCustody,
      RemandStatus: lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
    }
  }

  if (spiDefendant.CourtCorporateDefendant) {
    // Corporate Defendant
    const {
      Address: spiAddress,
      BailStatus: spiBailStatus,
      OrganisationName: { OrganisationName: spiOrganisationName },
      PNCidentifier: spiPNCidentifier
    } = spiDefendant.CourtCorporateDefendant

    return {
      Address: populateAddress(spiAddress),
      BailConditions: [],
      CourtPNCIdentifier: formatPncIdentifier(spiPNCidentifier),
      OrganisationName: spiOrganisationName,
      RemandStatus: lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
    }
  }

  throw new Error("CourtIndividualDefendant and CourtCorporateDefendant are missing from incoming record")
}

export default (courtResult: ResultedCaseMessageParsedXml): HearingDefendant => {
  const {
    Session: {
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  const { bailConditions, offences } = populateOffences(courtResult)

  const partialDefendant = processDefendant(spiDefendant)

  if (partialDefendant.BailConditions.length > 0) {
    partialDefendant.BailConditions = partialDefendant.BailConditions.concat(bailConditions)
  }

  const hearingDefendant: HearingDefendant = {
    ...partialDefendant,
    ArrestSummonsNumber: spiDefendant.ProsecutorReference,
    Offence: offences
  }

  return hearingDefendant
}
