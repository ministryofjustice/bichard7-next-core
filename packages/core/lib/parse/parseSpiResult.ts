import { XMLParser } from "fast-xml-parser"
import type { IncomingMessageParsedXml } from "../../types/SpiResult"
import { incomingMessageParsedXmlSchema } from "../../types/SpiResult"
import { decodeAttributeEntitiesProcessor, decodeTagEntitiesProcessor } from "../encoding"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeAttributeEntitiesProcessor,
    tagValueProcessor: decodeTagEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
