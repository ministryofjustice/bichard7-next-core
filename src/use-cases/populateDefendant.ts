import type { Address, DefendantDetail, HearingDefendant } from "src/types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml, SpiAddress, SpiCourtIndividualDefendant } from "src/types/IncomingMessage"
import { lookupRemandStatusBySpiCode } from "./dataLookup"
import PopulateOffences from "./PopulateOffences"

const formatPncIdentifier = (spiPNCIdentifier?: string): string | undefined =>
  spiPNCIdentifier ? spiPNCIdentifier.substring(0, 4) + "/" + spiPNCIdentifier.substring(4) : undefined

const populatePersonDefendantDetail = (spiCourtIndividualDefendant: SpiCourtIndividualDefendant): DefendantDetail => {
  const {
    PersonDefendant: {
      BasePersonDetails: {
        PersonName: {
          PersonTitle: spiPersonTitle,
          PersonGivenName1: spiPersonGivenName1,
          PersonGivenName2: spiPersonGivenName2,
          PersonGivenName3: spiPersonGivenName3,
          PersonFamilyName: spiPersonFamilyName
        },
        Gender: spiGender,
        Birthdate: spiBirthDate
      }
    }
  } = spiCourtIndividualDefendant

  const spiGivenNames = [spiPersonGivenName1, spiPersonGivenName2, spiPersonGivenName3]
    .filter((name) => !!name)
    .map((name) => name!.trim())

  return {
    PersonName: {
      Title: spiPersonTitle?.trim(),
      GivenName: spiGivenNames,
      FamilyName: spiPersonFamilyName
    },
    BirthDate: spiBirthDate,
    Gender: spiGender.toString()
  }
}

const populateAddress = (spiAddress: SpiAddress): Address => {
  if ("SimpleAddress" in spiAddress) {
    const {
      SimpleAddress: { AddressLine1, AddressLine2, AddressLine3, AddressLine4, AddressLine5 }
    } = spiAddress

    return {
      AddressLine1: AddressLine1,
      AddressLine2: AddressLine2,
      AddressLine3: AddressLine3,
      AddressLine4: AddressLine4,
      AddressLine5: AddressLine5
    }
  } else {
    const {
      ComplexAddress: { Locality, StreetDescription, Town, AdministrativeArea, UniqueStreetReferenceNumber }
    } = spiAddress

    const hearingOutcomeAddress = [
      UniqueStreetReferenceNumber?.toString(),
      StreetDescription,
      Locality,
      Town,
      AdministrativeArea
    ]
      .filter((x) => !!x)
      .reduce((acc: any, addressPart, index) => {
        acc[`AddressLine${index + 1}`] = addressPart!
        return acc
      }, {})

    return hearingOutcomeAddress
  }
}

export default (courtResult: ResultedCaseMessageParsedXml): HearingDefendant => {
  const hearingDefendant = {} as HearingDefendant
  const {
    Session: {
      Case: { Defendant: spiDefendant }
    }
  } = courtResult

  hearingDefendant.ArrestSummonsNumber = spiDefendant.ProsecutorReference

  if (spiDefendant.CourtIndividualDefendant) {
    const {
      CourtIndividualDefendant: {
        PersonDefendant: {
          PNCidentifier: spiPNCidentifier,
          BailConditions: spiBailConditions,
          ReasonForBailConditionsOrCustody: spiReasonForBailConditionsOrCustody
        },
        BailStatus: spiBailStatus,
        Address: spiAddress
      }
    } = spiDefendant
    hearingDefendant.CourtPNCIdentifier = formatPncIdentifier(spiPNCidentifier)
    hearingDefendant.DefendantDetail = populatePersonDefendantDetail(spiDefendant.CourtIndividualDefendant)
    hearingDefendant.Address = populateAddress(spiAddress)
    hearingDefendant.RemandStatus = lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
    hearingDefendant.BailConditions = spiBailConditions?.split(";").map((bailCondition) => bailCondition) || []
    hearingDefendant.ReasonForBailConditions = spiReasonForBailConditionsOrCustody
  } else if (spiDefendant.CourtCorporateDefendant) {
    // Corporate Defendant
    const {
      PNCidentifier: spiPNCidentifier,
      OrganisationName: { OrganisationName: spiOrganisationName },
      Address: spiAddress,
      BailStatus: spiBailStatus
    } = spiDefendant.CourtCorporateDefendant

    hearingDefendant.CourtPNCIdentifier = formatPncIdentifier(spiPNCidentifier)
    hearingDefendant.OrganisationName = spiOrganisationName
    hearingDefendant.Address = populateAddress(spiAddress)
    hearingDefendant.RemandStatus = lookupRemandStatusBySpiCode(spiBailStatus)?.cjsCode ?? spiBailStatus
  }

  const { offences, bailConditions } = new PopulateOffences(courtResult, hearingDefendant.BailConditions).execute()

  hearingDefendant.Offence = offences
  hearingDefendant.BailConditions = bailConditions

  return hearingDefendant
}
