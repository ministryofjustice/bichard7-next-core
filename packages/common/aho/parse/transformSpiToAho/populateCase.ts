import type { Case } from "../../../types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "../../../types/SpiResult"

import populateDefendant from "./populateDefendant"

export default (courtResult: ResultedCaseMessageParsedXml): Case => {
  const {
    Session: {
      Case: { PTIURN }
    }
  } = courtResult

  const ptiurn = PTIURN.toUpperCase()
  return {
    CourtReference: {
      MagistratesCourtReference: ptiurn
    },
    HearingDefendant: populateDefendant(courtResult),
    PreChargeDecisionIndicator: false,
    PTIURN: ptiurn
  }
}
