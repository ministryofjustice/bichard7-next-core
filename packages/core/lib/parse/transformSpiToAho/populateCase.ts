import type { ResultedCaseMessageParsedXml } from "../../../phase1/types/SpiResult"
import type { Case } from "../../../types/AnnotatedHearingOutcome"
import populateDefendant from "./populateDefendant"

export default (courtResult: ResultedCaseMessageParsedXml): Case => {
  const {
    Session: {
      Case: { PTIURN }
    }
  } = courtResult

  const ptiurn = PTIURN.toUpperCase()
  return {
    PTIURN: ptiurn,
    PreChargeDecisionIndicator: false,
    CourtReference: {
      MagistratesCourtReference: ptiurn
    },
    HearingDefendant: populateDefendant(courtResult)
  }
}
