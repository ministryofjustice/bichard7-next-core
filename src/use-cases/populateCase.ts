import type { Case } from "src/types/HearingOutcome"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import { createElement } from "src/types/XmlElement"
import populateDefendant from "./populateDefendant"

export default (courtResult: ResultedCaseMessageParsedXml): Case => {
  const {
    Session: {
      Case: { PTIURN }
    }
  } = courtResult

  return {
    PTIURN: createElement(PTIURN?.toUpperCase()),
    PreChargeDecisionIndicator: createElement("N", { Literal: "No" }),
    CourtReference: {
      MagistratesCourtReference: createElement(PTIURN)
    },
    HearingDefendant: populateDefendant(courtResult)
  }
}
