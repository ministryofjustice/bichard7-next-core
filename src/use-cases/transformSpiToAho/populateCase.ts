import type { Case } from "src/types/AnnotatedHearingOutcome"
import type { ResultedCaseMessageParsedXml } from "src/types/IncomingMessage"
import populateDefendant from "./populateDefendant"

export default (courtResult: ResultedCaseMessageParsedXml): Case => {
  const {
    Session: {
      Case: { PTIURN }
    }
  } = courtResult

  return {
    PTIURN: PTIURN?.toUpperCase(),
    PreChargeDecisionIndicator: false,
    CourtReference: {
      MagistratesCourtReference: PTIURN
    },
    HearingDefendant: populateDefendant(courtResult)
  }
}
