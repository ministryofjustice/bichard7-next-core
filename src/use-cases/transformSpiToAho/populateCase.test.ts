import { readFileSync } from "fs"
import type { HearingDefendant } from "src/types/AnnotatedHearingOutcome"
import parseSpiResult from "src/use-cases/parseSpiResult"
import populateCase from "./populateCase"
import populateDefendant from "./populateDefendant"

jest.mock("src/use-cases/transformSpiToAho/populateDefendant")

const message = readFileSync("test-data/input-message-001.xml", "utf-8")
const courtResult = parseSpiResult(message).DeliverRequest.Message.ResultedCaseMessage

describe("populateCase", () => {
  beforeAll(() => {
    const mockedPopulateDefendant = populateDefendant as jest.MockedFunction<typeof populateDefendant>
    mockedPopulateDefendant.mockReturnValue({} as HearingDefendant)
  })

  it("should transform SPI Case to Hearing Outcome Case", () => {
    const result = populateCase(courtResult)

    expect(result).toBeDefined()
    expect(result).toMatchSnapshot()
  })
})
