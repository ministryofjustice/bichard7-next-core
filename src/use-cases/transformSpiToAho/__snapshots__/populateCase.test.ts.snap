// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populateCase should transform SPI Case to Hearing Outcome Case 1`] = `
Object {
  "CourtReference": Object {
    "MagistratesCourtReference": "01ZD0303208",
  },
  "ForceOwner": Object {
    "BottomLevelCode": "",
    "OrganisationUnitCode": "0000000",
    "SecondLevelCode": "01",
    "ThirdLevelCode": "",
  },
  "HearingDefendant": Object {},
  "PTIURN": "01ZD0303208",
  "PreChargeDecisionIndicator": false,
}
`;
